File: frontend/src/api/accidentes.ts
import { get, post } from './client';
export type AccidenteEstado = 'ABIERTO' | 'EN_SEGUIMIENTO' | 'CERRADO';
export interface Accidente { id_accidente: string; persona_id: string; persona: string; cedula: string | null; area: string | null; fecha_accidente: string; clasificacion: string; estado: AccidenteEstado; descripcion: string; observacion: string | null; fecha_registro: string | null; origen: string; lote_id: string | null; lote_nombre_archivo: string | null; registrado_por: string | null; version: number; }
export interface LoteAccidente { id_lote: string; nombre_archivo: string; usuario_id: string; estado: string; total_filas: number; filas_validas: number; filas_con_error: number; filas_duplicadas: number; filas_importadas: number; fecha_creacion: string | null; }
export interface IncidenciaAccidente { id_error: string; numero_fila: number; codigo: string; mensaje: string; datos_fila: string | null; }
export const listarAccidentes = (params: URLSearchParams) => get<{items: Accidente[]; total: number; limite: number; offset: number}>(`/accidentes?${params}`);
export const obtenerAccidente = (id: string) => get<Accidente>(`/accidentes/${id}`);
export const crearAccidente = (data: {persona_id: string; fecha_accidente: string; clasificacion: string; descripcion: string; estado: AccidenteEstado; observacion: string | null}) => post<Accidente>('/accidentes', data);
export async function analizarArchivoAccidentes(file: File): Promise<{lote: LoteAccidente; puede_confirmarse: boolean; previsualizacion: unknown[]}> { const response = await fetch('/api/v1/importaciones/accidentes/analizar', { method: 'POST', credentials: 'include', body: (() => { const body = new FormData(); body.append('archivo', file); return body; })() }); if (!response.ok) throw new Error((await response.json()).message ?? 'No fue posible analizar el archivo.'); return response.json(); }
export const confirmarLoteAccidentes = (id: string) => post<LoteAccidente>(`/importaciones/accidentes/${id}/confirmar`, {});
export const listarLotesAccidentes = (params: URLSearchParams) => get<{items: LoteAccidente[]; total: number}>(`/importaciones/accidentes?${params}`);
export const listarIncidenciasAccidente = (id: string) => get<{items: IncidenciaAccidente[]; total: number}>(`/importaciones/accidentes/${id}/errores`);

File: backend/app/models/accidentes.py
from sqlalchemy import CheckConstraint, ForeignKey, Index, Integer, LargeBinary, String, UniqueConstraint
from sqlalchemy.orm import Mapped, mapped_column

from app.models.security import Base, MetadatosComunes


class Accidente(MetadatosComunes, Base):
    __tablename__ = "accidentes"
    __table_args__ = (
        UniqueConstraint("persona_id", "fecha_accidente", "clasificacion", name="uq_accidentes_persona_fecha_clasificacion"),
        CheckConstraint("version >= 1", name="version_positive"),
    )
    id_accidente: Mapped[str] = mapped_column(String, primary_key=True)
    persona_id: Mapped[str] = mapped_column(ForeignKey("personas.id_persona"), nullable=False, index=True)
    lote_id: Mapped[str | None] = mapped_column(ForeignKey("lotes_importacion_accidente.id_lote"), index=True)
    fecha_accidente: Mapped[str] = mapped_column(String, nullable=False)
    clasificacion: Mapped[str] = mapped_column(String, nullable=False)
    estado: Mapped[str] = mapped_column(String, nullable=False, default="ABIERTO")
    descripcion: Mapped[str] = mapped_column(String, nullable=False)
    observacion: Mapped[str | None] = mapped_column(String)


class LoteImportacionAccidente(MetadatosComunes, Base):
    __tablename__ = "lotes_importacion_accidente"
    __table_args__ = (CheckConstraint("version >= 1", name="version_positive"),)
    id_lote: Mapped[str] = mapped_column(String, primary_key=True)
    nombre_archivo: Mapped[str] = mapped_column(String, nullable=False)
    contenido_archivo: Mapped[bytes] = mapped_column(LargeBinary, nullable=False)
    usuario_id: Mapped[str] = mapped_column(ForeignKey("usuarios.id_usuario"), nullable=False, index=True)
    estado: Mapped[str] = mapped_column(String, nullable=False, default="ANALIZADO")
    total_filas: Mapped[int] = mapped_column(Integer, nullable=False)
    filas_validas: Mapped[int] = mapped_column(Integer, nullable=False)
    filas_con_error: Mapped[int] = mapped_column(Integer, nullable=False)
    filas_duplicadas: Mapped[int] = mapped_column(Integer, nullable=False)
    filas_importadas: Mapped[int] = mapped_column(Integer, nullable=False, default=0)


class ErrorImportacionAccidente(MetadatosComunes, Base):
    __tablename__ = "errores_importacion_accidente"
    __table_args__ = (CheckConstraint("version >= 1", name="version_positive"), Index("ix_errores_importacion_accidente_lote_fila", "lote_id", "numero_fila"))
    id_error: Mapped[str] = mapped_column(String, primary_key=True)
    lote_id: Mapped[str] = mapped_column(ForeignKey("lotes_importacion_accidente.id_lote"), nullable=False, index=True)
    numero_fila: Mapped[int] = mapped_column(Integer, nullable=False)
    codigo: Mapped[str] = mapped_column(String, nullable=False)
    mensaje: Mapped[str] = mapped_column(String, nullable=False)
    datos_fila: Mapped[str | None] = mapped_column(String)
